### DISCUSSION TITLE
use SSO rbac with claims other than groups

### DISCUSSION THREAD

**Discussion by sherlant** [argoproj/argo-workflows#12818] [Q&A] [answered]
Hi all,
i have read the doc about rbac : https://github.com/argoproj/argo-workflows/blob/main/docs/argo-server-sso.md
and i see that we can configure the rule annotation to map the subject of a user, and not a group claim.
But it doesn't work. I have try with :
```yaml
workflows.argoproj.io/rbac-rule: sub == 'tata'
```
anyone can help me to configure that ? I would like to have a service account per user in my keycloak idp.
thanks

**Comment by agilgur5** [comment]
~I believe you have to use `subject`, not `sub`.~

~The struct has it [named `Subject`](https://github.com/argoproj/argo-workflows/blob/87899e5dd66c73d14ef3f4acbfa25573d8cc3d4c/server/auth/sso/sso.go#L316) (that is [inherited](https://github.com/argoproj/argo-workflows/blob/87899e5dd66c73d14ef3f4acbfa25573d8cc3d4c/server/auth/types/claims.go#L12) from the `go-jose` JWT library).~

**EDIT**: nevermind, it gets [marshaled](https://github.com/go-jose/go-jose/blob/aabe78e07ec6844422c49167d53c8777a0e5b8c6/jwt/claims.go#L30) to `sub`

**Comment by sherlant** [comment]
Hello !
thanks for your response.
When i try to use subject, i have the following error : 
```
failed to evaluate rule: unknown name subject
```
I have try to use email instead, and it works. In user section in the argo-workflows UI, i see the subject of my user, and it's an ID, unusable in my context. The email is not perfect, i prefer to use the username of the user. Can we do that whith custom claim ? And add username in token scope in Keycloak ?

**Reply by agilgur5** [reply, marked as answer]
> When i try to use subject, i have the following error :

Sorry, that was my mistake, I hadn't looked at it in a bit. `go-jose` [marshals](https://github.com/go-jose/go-jose/blob/aabe78e07ec6844422c49167d53c8777a0e5b8c6/jwt/claims.go#L30) it to `sub`, so your original code was correct.

Your Server logs should have said it wasn't a match in the original version then.

> In user section in the argo-workflows UI, i see the subject of my user, and it's an ID, unusable in my context.

Yes this ID is what is assigned to `sub`.

> The email is not perfect, i prefer to use the username of the user. Can we do that whith custom claim ? And add username in token scope in Keycloak ?

You can only use the claims visible on that page. I added the `name` claim in #10927, so you could use that as a username (that is what the company I worked for at the time did)
